### ISSUE TITLE
Let's Sign - An Open Source Certificate Authority

### ISSUE THREAD

**Issue by ElTimuro** [documenso/backlog#21]
- To fully transform the signing ecosystem into an open economy, our long-term vision is to create a highly trusted Certificate Authority (CA), that dispenses free certificates for document signing, as Let's Encrypt does for SSL certificates
- To be a trusted entity that fulfills eIDAS and other regulatory requirements, Let's Sign can't be just a technical open source solution, but needs to be an independent, non-profit entity
- Open Source all used software and hardware will increase transparency in the space and help to keep the entity behind Let's SIgn lean by recruiting the whole signing ecosystem to participate.
- Let's Sign would be natively integrated into Documenso, giving adopters free certificates for their self-hosting efforts

**Comment by PC-SPEZIALIST** [comment]
**Let's Sign** sounds really awesome! 😍
More and more companies are interested in qualified electronic [signatures](https://roadmap.synaxon-services.de/p/neues-produkt-managed-unterschrift) based on [eIDAS](https://eidas.ec.europa.eu/efda/home).

**Comment by Dailson-Allves** [comment]
Allowing users to upload their own certificate would add a layer of trust, especially since A1-type certificates are recognized by all judicial authorities. This would be a highly valuable feature.

**Comment by zolfariot** [comment]
> Allowing users to upload their own certificate would add a layer of trust, especially since A1-type certificates are recognized by all judicial authorities. This would be a highly valuable feature.

With the current legal framework it would be impossible to obtain that level of flexibility and QES compliance at the same time (I speak for eIDAS in Europe, but it should be similar everywhere).
The certificates and private key chain-of-custody should be auditable by public entities at any time.
The two viable options for a project like Let's Sign would be either:

1) Store the private keys for each certificate in HSM modules hosted at Let's Sign, possibly with high-level hardware forced access control to enforce their usage only with the end-user verification (with a technology similar to the [one](https://www.nccgroup.com/media/2biaan4n/_final_public_report_ncc_group_google_encryptedbackup_2018-10-10_v10.pdf) used by Google Key Vault for encrypted android backups), end offer to the user a free qualified remote signing API (the user identify, send the document hash to Let's Sign, receive the signature).

2) Provide the private keys to the user in form of hardware tokens (smartcards, usb keys, on-premise HSM) - that can possibily be updated remotely to add/renew/remove certificates. The hardware token used have to be in specific whitelists defined by the regulator (EU for eIDAS QES signature).

**Comment by rht** [comment]
> The two viable options for a project like Let's Sign would be either:

I think option 1 is already what the documenso.com server does currently. The PDF signed by documenso.com has a cryptographic signature created by Documenso's own certificate. See the [pdf-sign](https://github.com/documenso/pdf-sign) library.

For option 2, I wrote something similar in https://github.com/documenso/backlog/issues/9#issuecomment-2869772233.

I think a more immediate goal for Let's Sign should be first to be an identity certificate dispenser, then become legally qualified later. The cryptographic signatures won't be as legally binding as a QES, but it is far better than the visual signatures. The analogy with Let's Encrypt is that just because a cert is issued by LE, doesn't mean you can trust the domain, but *at least* the transport protocol is HTTPS. We want cryptographic signatures to be the norm for digital documents. Also brewing: for digital media in general (which includes PDF), there is the [C2PA standard](https://c2pa.org/about/), that you can already find the signatures in images generated by OpenAI, several AI companies, and the hardware of several camera companies.

**Comment by scheibling** [comment]
Just stumbled over this software, have been working on this type of software in other contexts for quite a while now with the new requirements.


You technically don't have to become a trust service provider to ("help") provide AdES signatures (really, the qualified certificate isn't strictly necessary at this level either), since the user certificate management is already done by the respective local authorities and fulfills the requirements.

There are two paths that are generally used for this type of signature (Sweden and surrounding countries):

1 - Plain user signatures
Sign the document checksum with an eIDAS identity, and attach the data to the document (e.g. json/xml signature documents as attachments). This does kind of fulfill the first 1-2 steps in the archival requirements as well, but it won't show up as a "Green checkmark" in Adobe

2 - User + Seal
The user + seal approach is becoming the more common version. The user signatures are still created with the existing eIDAS eIDs, but the document is sealed (either after every signature, or once at the end) with a QES/AATL certificate. This provides the visual verification in Adobe, it effectively seals the document and all signatures in it and adds a time stamp and the validation information to the file, meaning it fulfills the requirements for LTA as well.

Both of these methods can be combined both with eIDAS and individual AES/QES certificates, this would require that either some form of central authority handles the eIDAS auth and then seals the document afterwards, or that the individual installations connect to eIDAS and get a QES certificate per instance for sealing documents after signing. We currently use #2 for one of our customers solutions.


P.s. One of the easier paths to being able to issue QT certificates in the long run would be to become a registration agent for one of the existing authorities. We have a registration agent for I.CA here in Sweden that's been delegated to complete the ID checks and issue QES certificates on I.CAs behalf. But the verification process does require about 2 hours of manual work per issued certificate for the required validation steps.


Edit: You might have done most of this already, just putting the info out there in case any of it is of any use

**Comment by zolfariot** [comment]
@scheibling I discovered recently that there are services offered by existing authorities that issue EUTL QES certificates with automated remote verification in a matter of 2-4 minutes (i.e., Evrotrust, Itsme for most EU country, plus many country-specific one).

WIth these services the certificate is remote and accessible through CSC-API or similar for remote signing (you send a document digest to the api and get back the signature). In some case the user has to pay upfront for 1-3 year certificate. In other cases the certificate is generated for free but then you pay something for each signature.

Already it would be very nice to be able to use these remote certificates within Documenso, but still the issuance will look much like in the old SSL days, manual and non-free.

In the long run having resources to run a foundation/non-profit that provides for free API for identity verification, QES certificate issuance (remote and/or locally on trusted crypto device), remote signature (for remote certificates) would really be the "Let's Encrypt"-style revolution in the space.
But similarly to how is not easy (and cheap) to pass all the scrutiny and audit needed to have your root CA listed by default in all browsers, it won't be easy for such a foundation to pass and maintain eIDAS/QES (or other regions') certification.

**Comment by scheibling** [comment]
@zolfariot In the long run that would be amazing, and I know some companies that would support an initiative like that at a post-concept stage (including mine). The pricing from a lot of the alternatives is absurdly high, and if this was priced lower or on the same level for organizations but free for individuals that could be a sustainable model in the long run.

But up until that point, like @rht says getting started with AdES(+Seal) might be an easier route - That effectively removes the current formal certification requirement, and is sufficient for like 95% of cases/documents at the moment. If you're not doing state/EU business or deliver payment solutions there aren't a lot of use cases for QC/QdES at the moment 

That way the only roadblock would be either A: finding a free/really cheap/sponsored way to automatically validate identities and issue certificates based on those, or B: finding a cheap way to connect to eIDAS and use the established EU-wide national IDs.

One verification that could be performed with minimal manual input is the new passport/intl. ID. They're certificate-based, and basically just require NFC+OCR and optionally a video biometric scan for validation. The latter part is more tricky though, for AdES the NFC+OCR might be enough but I'm far from sure on that.

We're currently starting work on a system in that direction for ID validation, so I'll know more about the practical/legal parts in a few weeks probably.

**Comment by koehn** [comment]
TL;DR: on the internet, nobody knows you’re a dog.

The challenge here is that while an HTTPS certificate is tied to a DNS name, an IV/OV certificate is tied to a non-digital thing (a person or organization). Proving a link between a certificate and a non-digital thing needs to be thorough and difficult/expensive to corrupt, which means it’s going to cost money.

I think that there is a market for non-OV/IV certificates that make a lesser claim to the identity of the source, using something like DNS to authenticate. It could be a simple CA like DKIM (where the public key is in DNS) or leverage the CA infrastructure that already exists like LetsEncrypt (better, since you get revocation). 

But it cannot provide the same assurance required to authenticate that a document was sealed by a person or organization. To do that requires bridging the analog and digital world, and that requires real work done by real people.

**Comment by masoudtahmasebi** [comment]
Hi dear creators, 
thank you for the great product, I wanted to ask when would Let's Sign be available to use in self hosted Documenso instances?

**Comment by scheibling** [comment]
So, after working with this for a while, and with a somewhat generous interpretation of the AdES standard, I think this could be accomplished with certain caveats.

In the EU, there is a (mostly) unified system for passport validation which is currently used by a lot of the larger identity verification systems. The user would need to provide a picture of the document + RFID scan to prove possession, and with that we can establish a unique identity and issue a short-term individual certificate. This method can be implemented with just the picture + RFID and a CRL check basically, meaning it shouldn't incur any significant costs.

This would potentially also require a picture submitted by the user which is compared to the passport, there are a couple of efficient local-only image models that should be able to do this with a reasonable amount of certainty quite efficiently. But since none of this has been legally challenged, it is unclear if this comparison actually has to be done on-the-fly pre-signature, or if the photos can be attached to the document/signature and compared manually in case a signature is challenged and needs to be validated later on.

This would require a server component and likely a mobile app component, for the RFID scans + live photo, but it's far from impossible to reach a sufficient standard while still maintaining the entire application as open source. It does still require that someone act as the third-party authority who is verifying these details, but with an open source model it can be distributed in a way so that multiple companies/organizations/nonprofits can implement it to give the user some options.

**Comment by koehn** [comment]
Leveraging existing government-validated digital ID (like EUID) is a feasible solution for IV certificates. We've looked at leveraging PKI-based driver's license standards (mDL, ISO 18013-5) and digital ID wallets like eiDAS 2.0 EUDI Wallets); all can provide reasonable proxies for positively identifying the individual. 

My guess is that governments will start supporting this eventually; probably the EU first. Cross-border interoperability will take a while from a legal perspective. 

We're working on a PDF validator product that allows you to automatically or through the browser (client-side) validate a PDF against a policy, allowing the user to express the CAs, PDF standards, etc. to which the PDF must comply. That way the customer need only specify the authorities and probably little else; we'll manage the details like CAs, revocation, conformance, detection of likely fraud, disposition, etc.

**Comment by masoudtahmasebi** [comment]
Hi and thank you for the great tool.
how come i combine documenso with https://github.com/LibPDF-js/core to achieve with eidas Level 2 - Advanced electronic signatures (AdES) ?
